src/main/java/pw/chew/transmuteit/DataManager.java

```diff
@@ -28,193 +19,201 @@
-            EconomyResponse r = econ.withdrawPlayer(player, balance);
-            EconomyResponse s = econ.depositPlayer(player, amount);
-        } else {
-            File userFile = new File(getDataFolder(), uuid.toString() + ".json");
-            try {
-                JSONObject data = getData(uuid);
-                data.put("emc", amount);
-                PrintWriter writer = new PrintWriter(userFile);
-                data.write(writer);
-                writer.close();
-            } catch(FileNotFoundException e) {
-                plugin.getLogger().severe("Unable to write to EMC file! EMC will NOT save!");
-            }
+        // If we're using Vault, use the vault API and take it from there.
+        if (useEconomy) {
+            econ.depositPlayer(player, amount - econ.getBalance(player));
+            return;
         }
+        // If not, use the inbuilt method.
+        getDataAndWrite(uuid, data -> data.put("emc", amount), "Setting EMC to " + amount);
     }
 
+    /**
+     * Resets a player's discoveries.
+     *
+     * @param uuid The UUID of the player.
+     */
     public void writeEmptyDiscovery(UUID uuid) {
-        File userFile = new File(getDataFolder(), uuid.toString() + ".json");
-        try {
-            JSONObject data = getData(uuid);
-            Map<String, Object> discoveries = new HashMap<>();
-            data.put("discoveries", discoveries);
-            PrintWriter writer = new PrintWriter(userFile);
-            data.write(writer);
-            writer.close();
-        } catch(FileNotFoundException e) {
-            plugin.getLogger().severe("Unable to write to EMC file! EMC will NOT save!");
-        }
+        getDataAndWrite(uuid, data -> data.put("discoveries", new HashMap<>()), "Emptying discoveries");
     }
 
+    /**
+     * Adds a new item to a player's discoveries.
+     *
+     * @param uuid The UUID of the player.
+     * @param item The item to be discovered.
+     */
     public void writeDiscovery(UUID uuid, String item) {
-        File userFile = new File(getDataFolder(), uuid.toString() + ".json");
-        try {
-            JSONObject data = getData(uuid);
-            data.getJSONArray("discoveries").put(item);
-            PrintWriter writer = new PrintWriter(userFile);
-            data.write(writer);
-            writer.close();
-        } catch(FileNotFoundException e) {
-            plugin.getLogger().severe("Unable to write to EMC file! EMC will NOT save!");
-        }
+        getDataAndWrite(uuid, data -> data.getJSONArray("discoveries").put(item), "Adding discovery " + item);
     }
 
+    /**
+     * Removes an item from a player's discoveries.
+     *
+     * @param uuid The UUID of the player.
+     * @param item The item to be removed.
+     */
     public void removeDiscovery(UUID uuid, String item) {
+        getDataAndWrite(uuid, data -> data.getJSONArray("discoveries").toList().remove(item), "Removing discovery " + item);
+    }
+
```

Comment:
```suggestion

    /**
     * Gets them writes specified data to a file.
     *
     * @param uuid The UUID of the player.
     * @param task The JSONObject to write
     * @param action The action performed, for debugging, in case of failure.
     */
```